src/ray/gcs/redis_context.cc

```diff
@@ -50,31 +57,41 @@ void SubscribeRedisCallback(void *c, void *r, void *privdata) {
   }
   int64_t callback_index = reinterpret_cast<int64_t>(privdata);
   redisReply *reply = reinterpret_cast<redisReply *>(r);
-  std::string data = "";
-  if (reply->type == REDIS_REPLY_ARRAY) {
-    // Parse the message.
+  std::vector<std::string> data;
+  // Parse the response.
+  switch (reply->type) {
+  case (REDIS_REPLY_ARRAY): {
+    // Parse the published message.
     redisReply *message_type = reply->element[0];
     if (strcmp(message_type->str, "subscribe") == 0) {
-      // If the message is for the initial subscription call, do not fill in
-      // data.
+      // If the message is for the initial subscription call, return the empty
+      // string as a response to signify that subscription was successful.
+      data.push_back("");
     } else if (strcmp(message_type->str, "message") == 0) {
       // If the message is from a PUBLISH, make sure the data is nonempty.
       redisReply *message = reply->element[reply->elements - 1];
-      data = std::string(message->str, message->len);
-      RAY_CHECK(!data.empty()) << "Empty message received on subscribe channel";
+      auto notification = std::string(message->str, message->len);
+      RAY_CHECK(!notification.empty()) << "Empty message received on subscribe channel";
+      data.push_back(notification);
     } else {
       RAY_LOG(FATAL) << "Fatal redis error during subscribe" << message_type->str;
     }
 
-    // NOTE(swang): We do not delete the callback after calling it since there
-    // may be more subscription messages.
-    RedisCallbackManager::instance().get(callback_index)(data);
-  } else if (reply->type == REDIS_REPLY_ERROR) {
+  } break;
+  case (REDIS_REPLY_ERROR): {
     RAY_LOG(ERROR) << "Redis error " << reply->str;
-  } else {
+  } break;
+  default:
     RAY_LOG(FATAL) << "Fatal redis error of type " << reply->type << " and with string "
                    << reply->str;
   }
+
+  if (callback_index >= 0) {
```

Comment:
Warrants an anon-namespace helper since it repeats, in my opinion
```
MaybeRemoveCallback(callback_index);
```
for
```
+  if (callback_index >= 0) {
+    bool delete_callback = RedisCallbackManager::instance().get(callback_index)(data);
+    if (delete_callback) {
+      RedisCallbackManager::instance().remove(callback_index);
+    }
+  }
```